Add `widths`, `heights` and `angles` setter to `EllipseCollection`

ericpre opened the pull request:
## PR summary

Currently, the following example:

```python
import matplotlib.pyplot as plt
from matplotlib.collections import EllipseCollection
import numpy as np

rng = np.random.default_rng(0)

widths = (2, )
heights = (3, )
angles = (45, )
offsets = rng.random((10, 2)) * 10

fig, ax = plt.subplots()

ec = EllipseCollection(
    widths=widths,
    heights=heights,
    angles=angles,
    offsets=offsets,
    units='x', 
    offset_transform=ax.transData,
    )

ax.add_collection(ec)
ax.set_xlim(-2, 12)
ax.set_ylim(-2, 12)

new_widths = rng.random((10, 2)) * 2
new_heights = rng.random((10, 2)) * 3
new_angles = rng.random((10, 2)) * 180

ec.set(widths=new_widths, heights=new_heights, angles=new_angles)
```
will fails with the errors:

```python
  File ~\Dev\matplotlib\lib\matplotlib\artist.py:147 in <lambda>
    cls.set = lambda self, **kwargs: Artist.set(self, **kwargs)

  File ~\Dev\matplotlib\lib\matplotlib\artist.py:1221 in set
    return self._internal_update(cbook.normalize_kwargs(kwargs, self))

  File ~\Dev\matplotlib\lib\matplotlib\artist.py:1213 in _internal_update
    return self._update_props(

  File ~\Dev\matplotlib\lib\matplotlib\artist.py:1187 in _update_props
    raise AttributeError(

AttributeError: EllipseCollection.set() got an unexpected keyword argument 'widths'
```

## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [n/a] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [n/a] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-examples-and-tutorials)
- [x] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [x] Documentation complies with [general](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/documenting_mpl.html#writing-docstrings) guidelines

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at https://matplotlib.org/devdocs/devel/development_workflow.html

- Create a separate branch for your changes and open the PR from this branch. Please avoid working on `main`.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  https://matplotlib.org/stable/devel/documenting_mpl.html#formatting-conventions.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

oscargus commented on lib/matplotlib/tests/test_collections.py:
```diff
@@ -398,6 +398,37 @@ def test_EllipseCollection():
+    new_angles = rng.random((10, 2)) * 180
+
+    ec.set(widths=new_widths, heights=new_heights, angles=new_angles)
+
```
Please add a check that the new values are actually set.

oscargus commented on lib/matplotlib/collections.py:
```diff
@@ -1813,6 +1813,18 @@ def _set_transforms(self):
             m[:2, 2:] = 0
             self.set_transform(_affine(m))
 
+    def set_widths(self, widths):
```
A basic doc-string for the added methods would be good.

oscargus reviewed (commented):
It probably makes sense to use the newly added methods in the constructor.

(Are there getters as well?)

oscargus reviewed (approved):
I cannot help with the getters, unfortunately. While it is indeed ideal to have both, it is clearly better to have only setters compared to nothing.

tacaswell commented:
We talked about this on the call and are 👍🏻 on it going in but would like getters.

The consensus on the call was for the getters to undo the geometric transformations, but don't worry about the ravel (if someone is passing in higher than 1D data, we would like to know why before bending over backwards to give it back).

ericpre commented:
I added the getters which return the same as what is given to the setters and rebased. The CI failures shouldn't be related to the changes of this PR.

timhoffm reviewed (changes requested):
In a similar case https://github.com/matplotlib/matplotlib/pull/26410#issuecomment-1910971749 where multiple related arrays can be set, we advocated for a central function that can take one or multiple of the arrays and make sure the final data (mixture of existing and new arrays) has consistent array lengths.

I suggest that we follow this approach here as well - at least for the geometry parameters *widths*, *heights*, *angles*, *offsets*. I grant that we won't check other potentially sequence-like parameters (e.g. linewidths), but AFAIK they are just cycled continuously and don't have to match in lengths.

QuLogic commented:
A friendly ping on this; I'd like to put out 3.9 beta/rc next week.

ericpre commented:
Thanks @QuLogic, yes it would be good to finish this PR. 

@timhoffm, does it sound good to overwrite the `set` method? In the example added in this PR, `set` is used these values in a single call. This has the advantage of keeping the API minimal and consistent.

timhoffm commented:
> @timhoffm, does it sound good to overwrite the `set` method?

That may work, but it's a bit tricky because the `set` methods are auto-generated to include all properties of the class:

https://github.com/matplotlib/matplotlib/blob/9e18a343fb58a2978a8e27df03190ed21c61c343/lib/matplotlib/artist.py#L139-L150

You may have to care for `_autogenerated_signature` and/or `_update_set_signature_and_docstring()` to get explicit `set` signature and docstring without manually having to specify all parameters. This is some you would have to fiddle into.

ericpre commented:
I added a `_check_length` method to check the length of the input parameter using `offsets` as reference for the number of items in the collection.

ericpre commented:
Similarly as in #26410, I remove checking the length of the parameters to leave for another PR.

QuLogic commented on lib/matplotlib/collections.py:
```diff
@@ -1795,6 +1795,42 @@ def _set_transforms(self):
             m[:2, 2:] = 0
             self.set_transform(_affine(m))
 
+    def _set_widths(self, widths):
+        self._widths = 0.5 * np.asarray(widths).ravel()
+
+    def _set_heights(self, heights):
+        self._heights = 0.5 * np.asarray(heights).ravel()
+
+    def _set_angles(self, angles):
+        self._angles = np.deg2rad(angles).ravel()
```
I'm not sure if we really need these private helpers.

ericpre replied:
I removed them.

QuLogic commented on lib/matplotlib/collections.py:
```diff
@@ -1795,6 +1795,42 @@ def _set_transforms(self):
+        self._set_widths(widths)
+        self.stale = True
+
+    def set_heights(self, heights):
+        """Set the lengths of second axes (e.g., minor axes)."""
+        self._set_heights(heights)
+        self.stale = True
+
+    def get_widths(self):
+        """Get the lengths of the first axes (e.g., major axis)."""
+        return self._widths * 2
+
```
These two are swapped order from the `.pyi` file.

QuLogic commented on lib/matplotlib/tests/test_collections.py:
```diff
@@ -408,6 +408,49 @@ def test_EllipseCollection():
+        offset_transform=ax.transData,
+        )
+
+    assert_array_almost_equal(ec._widths, np.array(widths).ravel() * 0.5)
+    assert_array_almost_equal(ec._heights, np.array(heights).ravel() * 0.5)
+    assert_array_almost_equal(ec._angles, np.deg2rad(angles).ravel())
```
Do we need to check internals now that we have the getters?

ericpre replied:
These check that the internal values are correct (multiplication of 0.5) and these are different from the values returned by the getter, so I would say that this is still worth checking these values?

timhoffm replied:
Testing internals is not ideal, but OTOH these tests do not impose a big liability. It's unlikely that we will refactor the internals and if so, the tests can be easily adapted.

Alternatives would be an image comparison (not favored) or checking the bounding box of an EllipseCollection with a single entry.

ericpre replied:
If this is find to keep to check as it is I would prefer that option because it is straighforward and more simple than the bounding box alternative.

ericpre reviewed (commented):
Thanks @QuLogic, I have addressed your comments!

timhoffm reviewed (approved):
Take or leave the internal check.

QuLogic commented:
I think we can ignore the AppVeyor failure here, as it seems to be something to do with wxPython.
